### PR TITLE
test_runner/performance: add sharded ingest benchmark

### LINKED ISSUE #9569: safekeeper: add WAL ingestion benchmarks with pageserver fan-out
For sharded ingest performance, we should add a benchmark that ingests WAL into a Safekeeper and fans it out into 8 shards across 8 pageservers.

Ideally, we would have a full end-to-end benchmark including compute, safekeeper cluster, and pageserver cluster with networking and S3. However, building this infrastructure is likely more work than we have resources for in this project.

For now, we'll write a Rust integration-style benchmark that runs the Safekeeper and Pageservers in memory on the local node, ignoring network latency, (possibly) disk latency, and compute overhead.

### PR DESCRIPTION
**Pull request by erikgrinaker**
## Problem

We need a benchmark to quantify the improvement of sharded ingestion.

Resolves #9569.

## Summary of changes

Adds a Python benchmark for sharded ingestion. This ingests 7 GB of WAL (100M rows) into a Safekeeper and fans out to 10 shards running on 10 different pageservers. The ingest volume and duration is recorded.

## Checklist before requesting a review

- [x] I have performed a self-review of my code.
- [ ] If it is a core feature, I have added thorough tests.
- [ ] Do we need to implement analytics? if so did you add the relevant metrics to the dashboard?
- [ ] If this PR requires public announcement, mark it with /release-notes label and add several sentences in this section.

## Checklist before merging

- [ ] Do not forget to reformat commit message to not include the above checklist

### CODE REVIEW

**Comment by erikgrinaker** [comment]
@VladLazar @jcsp Initial local results without sharded ingest:

```
MacBook Pro M3, 11 cores, 7 GB WAL (100M rows)

1 shard: 41.382s
2 shards: 40.805s
4 shards: 43.169s
10 shards: 61.436s
32 shards: 148.067s
```

Ingestion is IO bound with 1 shard, and pageserver ingestion is single-threaded so can use at most 1 CPU core. As long as pageservers have CPU headroom, the additional work due to sharding does not affect throughput and remains IO bound. Only when we saturate the CPU at ~10 shards does throughput drop, and then drops ~linearly up to 32 shards as expected.

This implies that we won't see ingestion throughput improvements unless pageservers see CPU exhaustion. We will see a reduction in CPU usage (and thus potential cost savings) though. However, we may see throughput increases if we saturate the bandwidth capacity of the Safekeepers too (bandwidth use currently scales linearly with shards). Just to set expectations for throughput improvements @jcsp.

**Comment by erikgrinaker** [comment]
Btw, writing this as a Rust in-memory benchmark wasn't practical due to the tight code coupling. A Pageserver WAL receiver pulls in most Pageserver infrastructure via the timeline/tenant, and all of this had to be made public for construction from benchmark binaries. We should add appropriate trait abstraction boundaries to allow constructing components in isolation for tests/benchmarks, but that's more work that it's worth right now.

File: test_runner/fixtures/neon_fixtures.py [hunk 1]
```diff
@@ -4710,17 +4710,13 @@ def tenant_get_shards(
...
-    else:
-        # Assume an unsharded tenant
-        return [(TenantShardId(tenant_id, 0, 0), override_pageserver or env.pageserver)]
+    return [
+        (
+            TenantShardId.parse(s["shard_id"]),
+            override_pageserver or env.get_pageserver(s["node_id"]),
+        )
+        for s in env.storage_controller.locate(tenant_id)
+    ]
```
[the remark this answers is not in the document]
Comment (erikgrinaker) [reply test_runner/fixtures/neon_fixtures.py]: I'll split this change out to a separate PR.

File: test_runner/performance/test_sharded_ingest.py [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+    neon_env_builder.num_pageservers = SHARD_COUNT
+    env = neon_env_builder.init_start()
+
+    # Create a sharded tenant and timeline, and migrate it to the respective pageservers.
+    tenant_id, timeline_id = env.create_tenant(shard_count=SHARD_COUNT)
+
+    for shard_number in range(0, SHARD_COUNT):
+        tenant_shard_id = TenantShardId(tenant_id, shard_number, SHARD_COUNT)
+        pageserver_id = shard_number + 1
+        env.storage_controller.tenant_shard_migrate(tenant_shard_id, pageserver_id)
```
Comment (VladLazar) [inline test_runner/performance/test_sharded_ingest.py:36]: This should happen on it's own, provided you call `storage_controller.reconcile_until_idle`. Still would be good to assert the following:
1. at the beginning of the test check that the shards are distributed correctly and stash the generation for each shard
2. at the end of the test, check that the generations are the same as in step 1 - this guarantees that no migrations have happened.

**Review by VladLazar — APPROVED** [review]

File: test_runner/performance/test_sharded_ingest.py [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+@pytest.mark.timeout(600)
+def test_sharded_ingest(neon_env_builder: NeonEnvBuilder, zenbenchmark: NeonBenchmarker):
+    """
+    Benchmarks sharded ingestion throughput, by ingesting a large amount of WAL into a Safekeeper
+    and fanning out to a large number of shards on dedicated Pageservers. Comparing the base case
+    (SHARD_COUNT=1) to the sharded case indicates the overhead of sharding.
+    """
+
+    # Specify the shard count and WAL volume. Each shard gets a dedicated pageserver.
+    SHARD_COUNT = 10
```
[the remark this answers is not in the document]
Comment (erikgrinaker) [reply test_runner/performance/test_sharded_ingest.py:24]: Good idea, done.

File: test_runner/performance/test_sharded_ingest.py (hunk 2 above)
Comment (erikgrinaker) [reply test_runner/performance/test_sharded_ingest.py]: Yeah, good call. For stuff like this I think it's better to explicitly specify the topology and disable any asynchronous actors, to ensure results are deterministic, but there doesn't seem to be a way to disable background reconciliation in the storage controller without also breaking foreground operations. I left a TODO to disable reconciliation and added assertions that shards don't move (the shard generation doesn't appear to be exposed via any HTTP APIs).

**Comment by erikgrinaker** [comment]
> Ingestion is IO bound with 1 shard (i.e. Safekeeper is bottleneck), and pageserver ingestion is single-threaded so can use at most 1 CPU core. As long as pageservers have CPU headroom, the additional work due to sharding does not affect throughput and remains IO bound (i.e. Safekeeper-bound).

This isn't entirely accurate. With 1 shard, the Safekeeper runs at 70% CPU while the Pageserver runs at 170% CPU -- no significant IO-wait. I think this indicates that we may get a 2x speedup with 2 shards, but it'll flatten out there (unless CPU is saturated). We'll find out once sharded ingest is implemented.

**Pull request merged**